tests/test_math_kvecgenerator.cc

```diff
@@ -0,0 +1,280 @@
+      BOOST_CHECK_LE(static_cast<double>(nvec), upperbound);
+      BOOST_CHECK_GE(static_cast<double>(nvec), lowerbound);
+    }  // for each cell
+  }    // end test case
+
+  /* ---------------------------------------------------------------------- */
+  /**
+   * @brief Test correct behavior under scaling
+   *
+   * If we double the cutoff as well as the basis vectors, the new
+   * vectors should be the doubled versions of the original vectors.
+   */
+  BOOST_FIXTURE_TEST_CASE(kvecgen_scaling_test, KvecgenRefFixture) {
+    // Get the required quantities for the tests
+    double cutoff = math::PI * 2;
+    double normsq;
+    for (auto & cell : this->cells) {
+      math::Kvectors kvectors(cutoff, cell, false);
+      int nvec = kvectors.get_numvectors();
+      Eigen::MatrixXd kvecs = kvectors.get_kvectors();
+      Eigen::VectorXd kvals = kvectors.get_norms();
+
+      // For multiple scaling factors, compare the k-vectors obtained
+      // before and after scaling. We are only using powers of two to
+      // eliminate unwanted behavior due to rounding errors.
+      double scales[2] = {0.5, 2};
+      for (auto & scale : scales) {
+        math::Kvectors kvectors_scaled(cutoff * scale, cell / scale);
+        int nvec_scaled = kvectors_scaled.get_numvectors();
+        Eigen::MatrixXd kvecs_scaled = kvectors_scaled.get_kvectors() / scale;
+        BOOST_CHECK_EQUAL(nvec, nvec_scaled);
+
+        // Check that the obtained vectors agree using squared norm
+        if (nvec == nvec_scaled) {
+          for (int ik = 0; ik < nvec; ik++) {
+            normsq = (kvecs.row(ik) - kvecs_scaled.row(ik)).squaredNorm();
+            BOOST_CHECK_LT(normsq, cutoff * cutoff * 1e-14);
+          }  // for all k-vectors (of this cell)
+        }    // if number of vectors agree
+      }      // for all scale factors
+    }        // for each cell
+  }          // end test case
+
+  /* ---------------------------------------------------------------------- */
+  /**
+   * @brief Test correct behavior under rotations
+   *
+   * If we rotate a (real space) cell and compute its k-vectors, the
+   * obtained k-vectors should be the same as the rotation applied to
+   * the k-vectors of the original cell (stored in the same order).
+   */
+  BOOST_FIXTURE_TEST_CASE(kvecgen_rotation_test, KvecgenRefFixture) {
+    // Get the required quantities for the tests
+    double cutoff = 8.0;
+    double normsq;
+    for (auto & cell : this->cells) {
+      math::Kvectors kvectors(cutoff, cell, false);
+      int nvec = kvectors.get_numvectors();
+      Eigen::MatrixXd kvecs = kvectors.get_kvectors();
+      Eigen::VectorXd kvals = kvectors.get_norms();
+
+      /* Define the used rotation matrices. In order to avoid
+       * discrepancies due to rounding errors, we only use rotations
+       * that are a combination of 90 degrees rotations, since these
+       * rotation matrices only contain 1, 0, and -1 as components.
```

Comment:
It might be more interesting to use arbitrary rotation, and check for approximate equality using epsilon ~1e-12. Such test would be able to catch more issues.

Reply:
I haven't seen any orthogonal matrix generator from Eigen, I think you could do

srand to set a fixed seed (as test hyperparameter, might need to include stdlib.h)
(according to eigen mail list https://listengine.tuxfamily.org/lists.tuxfamily.org/eigen/2010/07/msg00403.html)
and use Eigen's for each rotation rotation.setRandom() to make it random and use QR decomposition to get an orthogonal matrix as in
https://eigen.tuxfamily.org/dox/classEigen_1_1HouseholderQR.html#a3e8e56769bbaaed0616ad98c4ff99c7b

Reply:
Ok, so this test is a little more tricky than that. Before commit #9c310f3 , what @Luthaf suggests is what I had done as well , namely by explicitly generating some rotation matrices. However, if you want to perform the rotation test on realistic cells, you get into serious rounding error issues.
Most importantly, if you apply a rotation matrix, the number of k-vectors that are found within the cutoff is often not the same any more due to rounding errors. In that case, it makes no sense to check the equality of each pair of vectors, since there will be many vectors that only exist in the rotated cell or vice versa, and checking whether one vector was "left out" would also be a huge pain to make sure you don't fail due to edge cases.

There are some possible solutions to this problem. But before we call it a "problem" too much, I would like to remind you that we typically choose the cutoff to be so large that we don't really care about what happens at the boundary. So if some vector that have length "equal to" kcut aren't found due to these rounding errors, it doesn't really matter for applications.

This is why in commit #9c310f3, I have decided to change the rotation test to something numerically stabler, by choosing rotation matrices that only contain +1, -1 and 0's and thus preserve floating point arithmetic. (this is why in the scaling test, I only use factors of 0.5 and 2, by the way).
With the original rotation test with arbitrary matrices, I had a hand picked set of cells that worked out fine numerically, but I thought that this is not a good long term solution, if the failure of the tests depends too much on the floating point error you get from the specific implementation details (and I had the rotation test fail on me before because I changed something little).

Reply:
TLDR: Right now, we have a rotation test that is very stringent, i.e. it tests for essentially perfect equality between the results before and after rotation. Such tests are numerically challenging unless you use rotation matrices that do not affect floating point errors, which limits us to compositions of 90 degree rotations in xyz directions.

I did have some more alternatives in mind when I made these changes, which would come in the form of some additional tests for arbitrary rotations. If you want, I could add those. But given that the test file is already longer than both the kvec_generator.cc and .hh files (potentially longer than both of them combined once I move some of the functionality of the .hh file into the tests as suggested in this review), I'm not sure whether this would be more beneficial relative to the amount of extra code to maintain, since we have already 5 tests for a simple class that performs exactly one job.

Reply:
Does this problem occur for points exactly on the cutoff? Shouldn't this be fixable by adding a bit of numerical jitter to the cutoff? You use a cutoff which is a multiple of the cell parameters, only in such a case I can see that the double precision error of 1e-16 could change the number of k-vectors. What do you think

Reply:
It occurs for points close to the cutoff, yes. Back when I still had that in the test, I mostly used a cutoff that was not an integer multiple of the cell parameters, namely 6.28, and I still had these issues.
In principle, it wouldn't be difficult to add jitters to the cutoff until all tests pass, but that was exactly the kind of "fine-tuning" I was trying to avoid by rewriting the tests in the current form, since the tests suddenly started to fail when I changed something else in the code, although I don't recall which cell caused that.

Reply:
But don't you find it weird: One rotation can only introduce an error in the range 1e-16 (maybe 1e-15), so the k vector count can only change when a k vector is exactly on the cutoff. The probability that this happens is for a random cell should be super low. The basis vector are multiplied by pi so the basis vectors are not a multiple of your cutoffs. I don't understand how the count can change

Reply:
Yes, I agree. That's why I had printed all the vectors to see whether that was really the issue, and at least then, it was. I probably could have just shifted the cutoff by 1e-10 or so to get rid of that issue. The way I was generating the rotation matrices had huge potential for rounding errors, and combined with the loop over n3 that would accumulate those, I believe that errors on the order of 1e-14 might be realistic actually, but not much more.